Add Whole Word Masking and Padding Strategy to DataCollatorForLanguageModeling

rjgleaton opened the pull request:
Sorry about the previous PR. I made a git error when trying to amend my commits.

# What does this PR do?

Main contribution is adding Whole Word Masking to DataCollatorForLanguageModeling. This is to fix the issues mentioned in #12998 where the previous implementation only returned input_ids and labels. This implementation should pass through all values from the input mappings. Because I feel it's more intuitive and easier to maintain I've added an arg to DataCollatorForLanguageModeling that allows for whole word masking. The DataCollatorForWholeWordMasking collator has been largely deprecated and is now just a rapper for the language modeling collator where mlm and whole_word_mask are forced to true.

This implementation depends on offset mappings for determining if a given pair of tokens belong to the same word. As such, it only natively works for FastTokenizers. I'm not sure there's a tokenizer agnostic way to do this with non-fast tokenizers atm. 

~~Additionally I've adding an argument for specifying PaddingStrategy to DataCollatorForLanguageModeling. This is a minor QOL improvement that makes it more consistent with other collators.~~


Fixes #12998


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger, @Rocketknight1

Rocketknight1 reviewed (commented):
Hi @rjgleaton, thank you for the PR! We did some discussion internally and we think having a WWM masker / collator that works across lots of tokenizer classes would be great to have in the library.

I made a couple of comments, mainly about splitting this into two PRs (one for WWM, the other for adding `PaddingStrategy`) since the diff is already quite hard to review. Excited to get this in the library, though!

rjgleaton committed:
Add whole word masking

rjgleaton committed:
Vectorize whole word masking functions

rjgleaton committed:
Unit test whole word masking

rjgleaton commented:
This is just a quick callout - I added some unit tests for the two new functions I added since the logic isn't _super_ trivial. I'm not entirely clear on what conventions y'all prefer for creating unit tests, so please lmk if they need to be updated.

rjgleaton commented:
@Rocketknight1 pinging about this PR again. No rush, just making sure it doesn't get lost.

Rocketknight1 reviewed (approved):
Yes, I'm sorry for losing track of this! The GPT-OSS launch made me drop everything else for a week or two, lol. The new changes look good, but I made one comment. Sorry to keep iterating like this!

rjgleaton committed:
Remove support for TF in whole word masking

Rocketknight1 commented:
#40760 is merged! You can rebase this PR and ping me for a final review

rjgleaton commented:
> #40760 is merged! You can rebase this PR and ping me for a final review

@Rocketknight1 Should be good to review again. Some tests seem to be failing, though I don't _think_ they're related to my dev. Let me know if that's unexpected.

Rocketknight1 reviewed (approved):
Yes, this looks ready to merge now! Removing TF also cleaned up the diff a lot. Thank you for the nice PR, and for your patience with the delays and rebases!
